transformer/external/wasmtransformer.go

```diff
@@ -0,0 +1,242 @@
+	memData, err = mem.GetData(uint(directoryDetectOutputPointer), 8)
+	if err != nil {
+		return nil, err
+	}
+	resultPointer := binary.LittleEndian.Uint32(memData[:4])
+	resultLength := binary.LittleEndian.Uint32(memData[4:])
+	memData, err = mem.GetData(uint(resultPointer), uint(resultLength))
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debug(string(memData))
+
+	services := map[string][]transformertypes.Artifact{}
+	// will handles errors
+	err = json.Unmarshal(memData, &services)
+	return services, err
+}
+
+// Transform transforms the artifacts
+func (t *WASM) Transform(newArtifacts []transformertypes.Artifact, alreadySeenArtifacts []transformertypes.Artifact) ([]transformertypes.PathMapping, []transformertypes.Artifact, error) {
+	pathMappings := []transformertypes.PathMapping{}
+	createdArtifacts := []transformertypes.Artifact{}
+	data := make(map[string]interface{})
+	data["newArtifacts"] = newArtifacts
+	data["oldArtifacts"] = alreadySeenArtifacts
+	dataByt, err := json.Marshal(data)
+	if err != nil {
+		return nil, nil, err
+	}
+	dataStr := string(dataByt)
+
+	preopens := []string{}
+	for _, artifact := range newArtifacts {
+		for _, paths := range artifact.Paths {
+			for _, path := range paths {
+				preopens = append(preopens, path)
+			}
+		}
+	}
+
+	sort.Slice(preopens, func(i, j int) bool {
+		l1, l2 := len(preopens[i]), len(preopens[j])
+		if l1 != l2 {
+			return l1 < l2
+		}
+		return preopens[i] < preopens[j]
+	})
+
+	deduplicatedPreopens := []string{}
+	for _, path := range preopens {
+		shouldSkip := false
+		for _, existingPath := range deduplicatedPreopens {
+			if strings.HasPrefix(path, existingPath) {
+				shouldSkip = true
+				break
+			}
+		}
+
+		if !shouldSkip {
+			deduplicatedPreopens = append(deduplicatedPreopens, path)
+		}
+	}
+
+	finalPreopens := []string{}
+	for _, path := range deduplicatedPreopens {
+		finalPreopens = append(finalPreopens, path+":"+path)
+	}
+
+	vm, err := t.initVm(finalPreopens)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allocateResult, err := vm.Execute("malloc", int32(len(dataStr)+1))
+	if err != nil {
+		return nil, nil, err
+	}
+	dataPointer := allocateResult[0].(int32)
+	mod := vm.GetActiveModule()
+	mem := mod.FindMemory("memory")
+	memData, err := mem.GetData(uint(dataPointer), uint(len(dataStr)+1))
+	if err != nil {
+		return nil, nil, err
+	}
+	copy(memData, dataStr)
+	memData[len(dataStr)] = 0
+	transformOutput, err := vm.Execute("transform", dataPointer)
+	transformOutputPointer := transformOutput[0].(int32)
+	memData, err = mem.GetData(uint(transformOutputPointer), 8)
+	if err != nil {
+		return nil, nil, err
+	}
+	resultPointer := binary.LittleEndian.Uint32(memData[:4])
+	resultLength := binary.LittleEndian.Uint32(memData[4:])
+	memData, err = mem.GetData(uint(resultPointer), uint(resultLength))
+	if err != nil {
+		return nil, nil, err
+	}
+	logrus.Debug(string(memData))
+	var output transformertypes.TransformOutput
+	json.Unmarshal(memData, &output)
```

Comment:
Check the error returned by `json.Unmarshal`

Reply:
Resolved in d87520b